Improve tablet_metadata::clear_gently

bhalevy opened the pull request:
As seen in #23284, when the tablet_metadata contains many tables, even empty ones,
we're seeing a long queue of seastar tasks coming from the individual destruction of
`tablet_map_ptr = foreign_ptr<lw_shared_ptr<const tablet_map>>`.

This change improves `tablet_metadata::clear_gently` to destroy the `tablet_map_ptr` objects
on their owner shard by sorting them into vectors, per- owner shard.

Also, background call to clear_gently was added to `~token_metadata`, as it is destroyed
arbitrarily when automatic token_metadata_ptr variables go out of scope, so that the
contained tablet_metadata would be cleared gently.

Finally, a unit test was added to reproduce the `Too long queue accumulated for gossip` symptom
and verify that it is gone with this change.

Refs #23284

This change is not marked as fixing the issue since we still need to verify that there is no impact on query performance, reactor stalls, or large allocations, with a large number of tablet-based tables.

avikivity commented:
> 
> This change is not marked as fixing the issue since we still need to verify that there is no impact on query performance, reactor stalls, or large allocations, with a large number of tablet-based tables.

Why not verify it then?

avikivity commented on locator/token_metadata.cc:
```diff
@@ -1028,7 +1033,9 @@ token_metadata::clone_after_all_left() const noexcept {
 
 future<> token_metadata::clear_gently() noexcept {
-    return _impl->clear_gently();
+    if (auto impl = std::move(_impl)) {
```
std::exchange communicates the intent that you want to clear better.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -839,7 +839,12 @@ token_metadata::token_metadata(config cfg)
+token_metadata::~token_metadata() {
+    if (_impl) {
+        tlogger.debug("~token_metadata[{}]: need clear_gently, at {}", fmt::ptr(this), current_backtrace());
+        seastar::internal::run_in_background(clear_gently());
```
Perhaps we should formalize this facility of a shared_ptr gentle destroyer, like rcu. Instead of calling the destructor, link the object into a chain that will be called when possible (sort of like Linux rcu).

An advantage here is that the destructor becomes a few instructions instead of an out-of-line destructor call, so many compiler optimizations with shared pointer manipulations become possible. Think shared_sstable manipulations in containers.

I don't propose to do it now.

bhalevy replied:
> One way to do it: create a service in main.cc for destroying impls, pass the impl to it and wake the coroutine.

We can use the shared_token_metadata as this service.
If and when we implement a general background destruction service we can define something like that.
Also, why use a coroutine?  We can run a seastar fiber in the background that holds a gate, to be stopped in stop().

> Advantage: can run in maintenance service level.

Since clear_gently is freeing memory I think that the memtable_scheduling_group is more appropriate as it will get more shares when memory is low.

avikivity replied:
memtable_scheduling_group shares aren't related to general free memory.

avikivity replied:
No

bhalevy replied:
used maintenance_scheduling_group in v2

avikivity replied:
Those things are terrible, let's use closed loops where we can.

bhalevy commented:
In v2 ([21a634f](https://github.com/scylladb/scylladb/commit/21a634fb2bc782ea2883bc7ded935619cea3cbaa)):
- rewritten the background disposal part

bhalevy commented:
In v3 ([0deb1f8](https://github.com/scylladb/scylladb/commit/0deb1f85d2c5960951ba8f096a4de3b007531746)):
- Fixed #includes using double quotes
- Skipped unit test in release mode (due to https://github.com/scylladb/scylladb/issues/24641)
- `token_metadata: release version before clear_and_dispose` to make sure the topology versions are released as soon as possible while the rest is cleared gently.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -1216,7 +1216,7 @@ future<> shared_token_metadata::mutate_on_all_shards(sharded<shared_token_metada
     std::vector<mutable_token_metadata_ptr> pending_token_metadata_ptr;
     pending_token_metadata_ptr.resize(smp::count);
-    auto tmptr = make_token_metadata_ptr(co_await stm.local().get()->clone_async());
+    auto tmptr = stm.local().make_token_metadata_ptr(co_await stm.local().get()->clone_async());
```
Very good, I like making dependencies explicit like this.

avikivity commented on utils/background_disposer.cc:
```diff
@@ -0,0 +1,92 @@
+}
+
+background_disposer::~background_disposer() {
+    SCYLLA_ASSERT(_stopped);
```
???

avikivity commented on utils/background_disposer.cc:
```diff
@@ -0,0 +1,92 @@
+    }
+    logger.debug("dispose: queuing item={} for disposal", fmt::ptr(item.get()));
+    try {
+        _items.push_back(std::move(item));
```
Alternative 1: the item is always a member of a list, either a live list or a to-be-disposed list, and you use list::splice to move from one list to another, no allocations needed.
Alternative 2: intrusive list.

bhalevy replied:
Intrusive list was my first try, but as you said, we need to keep the object alive so the disposer becomes a factory for the objects with life-cycle management.
It's possible and could be used both for unique ptrs and shard ptrs, but it's more complicated.

avikivity replied:
Then alternative 1 should work.

But maybe synchronous destruction on exception isn't too bad, it's unlikely to ever happen.

bhalevy replied:
used intrusive_list in v4

avikivity commented on utils/background_disposer.hh:
```diff
@@ -0,0 +1,55 @@
+    std::list<std::unique_ptr<basic_item>> _items;
+public:
+    background_disposer();
+    background_disposer(background_disposer&&);
```
noexcept? typically important for move constructors.

bhalevy commented:
> #### Failed Tests (4/1047):
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.1.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.2.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.3.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.4.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)

I don't understand what happened as the log files are empty.

Cc @yaronkaikov

Annamikhlin commented:
> > #### Failed Tests (4/1047):
> > 
> > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.1.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.2.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.3.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.4.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> 
> I don't understand what happened as the log files are empty.
> 
> Cc @yaronkaikov

`Cannot contact spider4.cloudius-systems.com`
Unfortunately, we experienced several power outages in the office today, which affected connectivity to local servers

bhalevy commented:
> > > #### Failed Tests (4/1047):
> > > 
> > > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.1.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.2.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.3.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17909/testReport/junit/cluster.test_tablets_merge.4.xunit/xml/debug___Unit_Tests_Custom___debug____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> > 
> > 
> > I don't understand what happened as the log files are empty.
> > Cc @yaronkaikov
> 
> `Cannot contact spider4.cloudius-systems.com` Unfortunately, we experienced several power outages in the office today, which affected connectivity to local servers

It would be good to somehow reflect that on Jenkins and/or in the automated reports on github.

bhalevy commented:
In v4 ([009792a](https://github.com/scylladb/scylladb/commit/009792a05ec86db7c73dd3ca72d12786c0cbacb6)):
- marked `background_disposer(background_disposer&&)` and `basic_item(basic_item&&)` as `noexcept`
- Used intrusive list in background_disposer
- Simplified the disposer() loop implementation
- Improved assertion in ~bacggrund_disposer to allow for early destroy, as long as the _items list is empty.

tgrabiec commented on locator/token_metadata.cc:
```diff
@@ -1035,8 +1037,19 @@ token_metadata::clone_after_all_left() const noexcept {
 
+void token_metadata::clear_and_dispose_impl() noexcept {
+    if (!_shared_token_metadata) {
+        return;
```
Shouldn't we _impl->clear_gently() in that case in token_metadata::clear_gently()?

tgrabiec commented on test/cluster/test_tablets_merge.py:
```diff
@@ -388,6 +388,71 @@ async def finished_merging():
+    insert = cql.prepare(f"INSERT INTO {ks}.test(pk, c) VALUES(?, ?)")
+    for pk in keys:
+        value = random.randbytes(2000)
+        cql.execute(insert, [pk, value])
```
Could be parallelized with asyncio.gather().

tgrabiec commented on test/cluster/test_tablets_merge.py:
```diff
@@ -388,6 +388,71 @@ async def finished_merging():
+    keys = range(total_keys)
+    insert = cql.prepare(f"INSERT INTO {ks}.test(pk, c) VALUES(?, ?)")
+    for pk in keys:
+        value = random.randbytes(2000)
```
could share the value, I think

bhalevy replied:
Right, copy-pasta... the value doesn't need to be randomized since sstable_compression is disabled.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -335,17 +335,20 @@ void tablet_metadata::drop_tablet_map(table_id id) {
+    // are owned by this shard, and releases them.
+    // Releasing the foreign lw_shared_ptr does not cause reallocation of
+    // the _tablets map, so we can safely iterate over it.
+    co_await smp::invoke_on_all([this] -> future<> {
```
Actually, it is probably called on other shards.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | i-05388e511729808de |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/)<br> :small_blue_diamond: boost/background_disposer_test<br> :small_blue_diamond: boost/storage_proxy_test<br> :small_blue_diamond: boost/tablets_test<br> :small_blue_diamond: boost/token_metadata_test<br> :small_blue_diamond: cluster/test_tablets_merge | :white_check_mark: | spider1.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/)<br> :small_blue_diamond: boost/background_disposer_test<br> :small_blue_diamond: boost/storage_proxy_test<br> :small_blue_diamond: boost/tablets_test<br> :small_blue_diamond: boost/token_metadata_test<br> :small_blue_diamond: cluster/test_tablets_merge | :white_check_mark: | spider8.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17917/)<br> :small_blue_diamond: boost/background_disposer_test<br> :small_blue_diamond: boost/storage_proxy_test<br> :small_blue_diamond: boost/tablets_test<br> :small_blue_diamond: boost/token_metadata_test<br> :small_blue_diamond: cluster/test_tablets_merge | :white_check_mark: | spider2.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5715/) | :x: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8904/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16999/) | :white_check_mark: |  |

#### Failed Tests (1/8663):
- [test_streaming_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17917/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/release___dtest_with_tablets___test_streaming_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_streaming_decommission)

#### Build Details:
- Duration: 3 hr 11 min

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -1035,8 +1037,19 @@ token_metadata::clone_after_all_left() const noexcept {
+    }
+    if (auto impl = std::exchange(_impl, nullptr)) {
+        _shared_token_metadata->clear_and_dispose(std::move(impl));
+        _shared_token_metadata = nullptr;
```
> Is it guaranteed that no token_metadata_ptr outlives shared_token_metadata? The pointers are captured by lots of things, also indirectly via erm.

shared_token_metadata is started very early, right after the snitch, and therefore should be destroyed almost last, and in particular, after the erm_factory (Right now that code is commented out in main because of dependency on storage_proxy that indeed holds, see 86296ba557fc8e1a9d059fe6fd1b3f66877d98c8, https://github.com/scylladb/scylladb/pull/22017, and https://github.com/scylladb/scylladb/issues/22886)

> Why not just clear_gently() in the background? Clearing is always faster than constructing, so those fibers should not accumulate.

This was pretty much my first attempt, using seastar::run_in_background.
Can also achieve that with a gate to be closed in shared_token_metadata::stop() (second try)
But that can create a long queue of tasks if there's a burst of disposals so it'd be better to explicitly manage that queue per shard.  Parallelizing  many clear_gently fibers on the same shard doesn't buy us much (unless they invoke_on other shards like the tablet_metadata clear_gently does today or do I/O which is less likely).

To with the lifecycle management, we can have the background_disposer almost manufacture the objects and provide a std::unique_ptr<T, background_disposer::delete_gently> that will destroy the object in the background.
Similarly we can extend it in the future also for shared pointers that will be disposed in the background when they lose their last reference.

bhalevy commented on test/cluster/test_tablets_merge.py:
```diff
@@ -388,6 +388,71 @@ async def finished_merging():
+    async def check_logs(when):
+        for server in servers:
+            log = await manager.server_open_log(server.server_id)
+            matches = await log.grep("Too long queue accumulated for gossip")
```
What do you suggest?
Unfortunately, with the lowered threshold I also see similar warnings for the main scheduling group which are unrelated to token metadata.

bhalevy commented:
In v5 ([25b52f2](https://github.com/scylladb/scylladb/commit/25b52f22d63ec5fd5e5b54699027a5cb7ba24e05)):
- simplified, using a gate in shared_token_metadata to await background clear_gently fibers to facilitate easier backporting
- dropped introduction of a background_disposer class which isn't needed by the simplified design.
- dropped scheduling group patch.  I think it requires more research under memory pressure.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17957/) | :white_check_mark: | i-09fb1d58428fd06a8 |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17957/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17957/) | :white_check_mark: | spider4.cloudius-systems.com |

#### Build Details:
- Duration: 1 hr 6 min

bhalevy commented:
> > > > This change is not marked as fixing the issue since we still need to verify that there is no impact on query performance, reactor stalls, or large allocations, with a large number of tablet-based tables.
> > > 
> > > 
> > > Why not verify it then?
> > 
> > 
> > We need a SCT test for that
> 
> And...?

@vponomaryov can you please run the scale-5000-tables-test against this branch to see if it fixed https://github.com/scylladb/scylladb/issues/23284 entirely?

bhalevy commented:
> > > > This change is not marked as fixing the issue since we still need to verify that there is no impact on query performance, reactor stalls, or large allocations, with a large number of tablet-based tables.
> > > 
> > > 
> > > Why not verify it then?
> > 
> > 
> > We need a SCT test for that
> 
> And...?

@avikivity should SCT block merging of this PR?
The unit test included here already shows some improvement, but even if it's insufficient, we can continue the work as a followup.

avikivity commented:
> > > > > This change is not marked as fixing the issue since we still need to verify that there is no impact on query performance, reactor stalls, or large allocations, with a large number of tablet-based tables.
> > > > 
> > > > 
> > > > Why not verify it then?
> > > 
> > > 
> > > We need a SCT test for that
> > 
> > 
> > And...?
> 
> @avikivity should SCT block merging of this PR? The unit test included here already shows some improvement, but even if it's insufficient, we can continue the work as a followup.

It's good to have more realistic testing.

In any case, we did not get a full CI run.

bhalevy commented:
> > So, the scenario that worked on vnodes using SI in addition still doesn't work with tablets.
> 
> As I thought, this PR made incremental improvement allowing us to get to the next roadblock, so it can't be considered as a fix to #23284. We can create a sub-issue for the particular problem fixed here, and open a sub-issue for
> 
> ```
> scylla[5697]:  [shard  0:strm] seastar - (rate limiting dropped 50 similar messages) Too long queue accumulated for streaming (5041 tasks)
>                                                                               5011: N7seastar8internal21coroutine_traits_baseIvE12promise_typeE
> ```

@vponomaryov I opened https://github.com/scylladb/scylladb/issues/24818 for the next long task queue, apparently coming from repair.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | ran |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/)<br> :small_blue_diamond: test/boost/storage_proxy_test.cc<br> :small_blue_diamond: test/boost/tablets_test.cc<br> :small_blue_diamond: test/boost/token_metadata_test.cc<br> :small_blue_diamond: cluster/test_tablets_merge | :white_check_mark: | spider3.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/)<br> :small_blue_diamond: test/boost/storage_proxy_test.cc<br> :small_blue_diamond: test/boost/tablets_test.cc<br> :small_blue_diamond: test/boost/token_metadata_test.cc<br> :small_blue_diamond: cluster/test_tablets_merge | :white_check_mark: | spider7.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18102/)<br> :small_blue_diamond: test/boost/storage_proxy_test.cc<br> :small_blue_diamond: test/boost/tablets_test.cc<br> :small_blue_diamond: test/boost/token_metadata_test.cc<br> :small_blue_diamond: cluster/test_tablets_merge | :white_check_mark: | spider8.cloudius-systems.com |
| dtest | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9090/) | :white_check_mark: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5864/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17188/) | :white_check_mark: |  |

#### Build Details:
- Duration: 8 hr 35 min

bhalevy commented:
@scylladb/scylla-maint please consider merging

bhalevy commented:
In v6 ([6e4803a](https://github.com/scylladb/scylladb/commit/6e4803a750e42ea7504ec7f69f2a4ed02782b3e4)):
- locator: tablets: tablet_metadata: clear_gently: optimize foreign ptr destruction
  - pre-sort the foreign table map ptrs to make their destruction loop on each shard more efficient
- token_metadata: clear_and_destroy_impl when destroyed
  - do not destroy _impl in token_metedata:clear_gently - it should be valid after clear_gently, as some unit tests expect it.
  - added missing deferred  deferred_stop(shared_token_metadata) in unit tests that failed on exit when the gate wasn't closed.

test.py passed locally in dev mode
